code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+In contrast static type checking does not need to worry about performance (as much). Afterall, this style of type checking happens in the developer's time (hence the "as much"), and not the end user's time. This enables a static type checker to do more complex type of checks and give better hints as to what is going wrong. However, to perform static type checking the information on types needs to be available before running the code. That often means that you as a programmer need to add this information. That is both a blessing and a curse. You will have to write more code, but the added information will make it easier for your co-workers or your future self to understand.
+
+Static type checking is often preferred, and so much so that languages such as JavaScript (in the form of TypeScript) and Python have started to adopt type information to enable static type checking. 
+</details>
+
+
+## Type hints in Python
+
+<details markdown="1"><summary  markdown="span">Type hints</summary>
+A type hint in the simplest form looks like this:
+
+
+    foo: int
+
+
+All this says is, the type of the variable `foo` should be an integer. Notice how there is no initial value here. This line of code does not create a variable `foo`, all it does is add a hint that `foo`, once it exists, should be an integer. That means this will raise a `NameError`:
+
+
+    $ python
+    >>> foo: int
+    >>> foo
+    NameError: name 'foo' is not defined
+
+
+It is possible to combine type hints and initialization on the same line, like so:
+
+
+    foo: int = 3
+
+
+That looks somewhat redundant, doesn't it? How can the *literal* `3` be anything else than an integer? This is where type inference kicks in. Tools such as `mypy` will try to infer the types of variables from their use. It is quite safe to assume type inference is possible here, so probably best to just write:
+
+
+    foo = 3
+
+
+Type inference does have its limitations, for instance `mypy` will not do any type inference in functions without type hints. To understand why, let's quickly look into function type hints. In the simplest form:
+
+
+    def add(a: int, b: int) -> int:
+        c = a + b
+        return c
+
+
+The syntax is relatively straight forward, using the colon (`:`) for parameter type hints, and the arrow (`->`) for the return type. Notice how the type of `c` is not annotated. It can be, but it is not needed. From the types of `a` and `b` and the `+` operation, `mypy` can infer the type of `c`. But what if we did not annotate this function. Well, in that case, `a` and `b` could be anything: `str`, `float`, `list`, you name it! This is where `mypy` draws a line, if you do not annotate a function, `mypy` will not even attempt to do type inference. Instead all variables will be of type `Any`.
+
+What is `Any`? Well, anything really. It is an escape hatch of sorts that provides no information. Once `Any` gets involved type checking becomes rather impossible. What is `Any + int`? `Any`
```

Comment:
Hoe relevant is het om `Any` te benoemen? Gaan ze dit tegenkomen in de output van Mypy ofzo?

Reply:
Ja heel vaak als code maar deels ge-annoteerd is. Als je nu met je muis over Python code heengaat in VSC, zul je ook veel Any tegenkomen